ProcessOptimizer/space/constraints.py

```diff
@@ -121,6 +124,100 @@ def rvs(self, n_samples=1, random_state=None):
         # We draw more samples when needed so we only return n_samples of the
         # samples
         return rows[:n_samples]
+    
+    def sumequal_sampling(self, n_samples=1, random_state=None):
+        """Draw samples that respect sum_equals constraints.
+
+        The samples are in the original space. They need to be transformed
+        before being passed to a model or minimizer by `space.transform()`.
+
+        Parameters
+        ----------
+        * `n_samples` [int, default=1]:
+            Number of samples to be drawn from the space.
+
+        * `random_state` [int, RandomState instance, or None (default)]:
+            Set random state to something other than None for reproducible
+            results.
+
+        Returns
+        -------
+        * `points`: [list of lists, shape=(n_points, n_dims)]
+           Points sampled from the space.
+        """
+
+        rng = check_random_state(random_state)
+                
+        # Find a point on the plane defined by the sum_equals constraint where
+        # A + B + ... = value. We do this by asking where the diagonal between
+        # the origin and A_max, B_max, ... intersects the plane.
+        d = len(self.sum_equals[0].dimensions)
+        origin = np.zeros(d)
+        delta = np.zeros(d)        
+        for i, dim in enumerate(self.sum_equals[0].dimensions):
+            origin[i] = self.space.bounds[dim][0]
+            delta[i] = self.space.bounds[dim][1] - self.space.bounds[dim][0]
+        A = np.zeros((d,d))
+        B = np.zeros(d)
+        # Row representing the sum constraint
+        A[0,:] = 1
+        B[0] = self.sum_equals[0].value
+        # Rows that define the linear equation for the diagonal along the 
+        # constrained dimensions
+        for i in range(1,d):
+            A[i, i-1] = -delta[i]/delta[0]
```

Comment:
```suggestion
            A[i, i-1] = -delta[i]/delta[i-1]
```

Reply:
I cannot seem to reproduce this error. I tried building the exact space you give above and ran 10,000 points without getting anything outside the space:

```
constraints = Constraints([SumEquals([1,2,0], 10)], space=[(1.0,5), (1.0, 10), (1.0, 100)],)

for i in range(10000):
    x = constraints.sumequal_sampling()
    x = x[0] # To handle that we get a list of lists
    in_space = np.all([x[i] > constraints.space.bounds[i][0] 
            and x[i] <= constraints.space.bounds[i][1] 
            for i in range(len(x))])
    if not in_space:
        break
```

Reply:
Argh, it's unfair to edit your reply when I'm trying to respond to it :P

Reply:
The `point` I mention that can be outside the space is not the points found, but the solution to the linear equations. But this is why that is a problem.

Reply:
I decided to re-do the derivation (because of course I threw away my notes from the first time...), so could you check that this formulation of the linear algebra is correct for any dimension?
![20230322_093957971_iOS](https://user-images.githubusercontent.com/64902587/226863087-acf152d1-aae5-444f-acf8-91a6625db2fd.jpg)
The fix involves making use of the origin, which as you have pointed out is entirely missing from the code, but should influence the solution. Turns out that both A and B need fixing :P

Reply:
That is a way better way to handle the origin than my suggestion.

Sorry for the many edits, the matrix formalism of linear algebra is really not my strong suit, so I make so many errors that I then have to fix.

Reply:
No worries, I only got through it this quickly because I did this derivation (and clearly got it wrong) when I wrote it the first time, where this derivation took me a good half day to figure out...

Reply:
I've pushed the new implementation now. This is a good demonstration of the need for tests, so I will move on to that next. I will mark this particular thread as resolved.